File: solutions/3958-minimum-removals-to-balance-array/minimum-removals-to-balance-array.java
import java.util.Arrays;

class Solution {
    public int minRemoval(int[] nums, int k) {
        Arrays.sort(nums); // Sort array to maintain order
        int n = nums.length;
        int maxSize = 0;
        int left = 0;

        // Use sliding window to find longest valid subarray
        for (int right = 0; right < n; right++) {
            // While condition is violated, shrink window from left
            while (nums[right] > (long) k * nums[left]) left++;
            maxSize = Math.max(maxSize, right - left + 1);
        }
        // Return number of elements to remove
        return n - maxSize;
    }
}

File: solutions/3958-minimum-removals-to-balance-array/minimum-removals-to-balance-array.js
var minRemoval = function(nums, k) {
    nums.sort((a, b) => a - b);
    let left = 0, maxSize = 0;

    for (let right = 0; right < nums.length; right++) {
        while (nums[right] > k * nums[left]) left++;
        maxSize = Math.max(maxSize, right - left + 1);
    }
    return nums.length - maxSize;
};